Added windows and uwp platform topic pages (#77)
microsoft/vcpkg-docs: vcpkg/users/platforms/windows.md

        target_link_options(mytarget PRIVATE /CETCOMPAT)
    endif()
endif()
```

* For improved Standard C/C++ Conformance, use the latest switch settings.

```
if(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
    target_compile_options(mytarget PRIVATE /permissive- /Zc:__cplusplus /Zc:inline)

    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 19.26)
        target_compile_options(mytarget PRIVATE /Zc:preprocessor)
    endif()

    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 19.28)
        target_compile_options(mytarget PRIVATE /Zc:lambda)
    endif()

    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 19.35)
        target_compile_options(mytarget PRIVATE /Zc:templateScope)
    endif()
endif()
```

* To support the use of Whole Program Optimization / Link-Time Code Generation, recommended build settings are as follows:

```
if((CMAKE_CXX_COMPILER_ID MATCHES "MSVC") AND CMAKE_INTERPROCEDURAL_OPTIMIZATION)
    target_compile_options(${PROJECT_NAME} PRIVATE /Gy /Gw)

    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 19.35)
        target_compile_options(mytarget PRIVATE /Zc:checkGwOdr)
    endif()
endif()
```

> * If enabling Spectre migations, use the following guards.

Reviewer: ```suggestion
* If enabling Spectre mitigations, use the following guards.
```

---
title: Microsoft Visual C++ for Windows
description: Building for Windows with the Microsoft Visual C++ Compiler
ms.date: 04/28/2023
---
# Microsoft Visual C++

## Triplets

> vcpkg includes [triplets](https://github.com/microsoft/vcpkg/tree/master/triplets) for building Windows desktop applications using the MSVC ``cl.exe`` compiler.

Reviewer 2: See my suggestion for uwp.

There are also a number of [community triplets](https://github.com/microsoft/vcpkg/tree/master/triplets/community) for the MSVC compiler toolset as well.

| Architecture | vcpkg community triplets |
|--------------|--------------------------|
| x64          | x64-windows-static-md    |
| x86          | x86-windows-static       |
|              | x86-windows-static-md    |
| arm          | arm-windows              |
|              | arm-windows-static       |
| arm64        | arm64-windows-static     |
|              | arm64-windows-static-md  |
|              | arm64ec-windows          |

> [!NOTE]
> Community triplets are not tested as part of vcpkg repository's CI process, so regressions can occur as part of library updates. PRs improving support are welcome!

The ``static`` linking triplets are set to use the MSVC Runtime as a static library (i.e. ``VCPKG_CRT_LINKAGE static``).

The ``static-md`` linking triplets are set to use the MSVC Runtime as a DLL (i.e. ``VCPKG_CRT_LINKAGE dynamic``). This is the recommended solution for redistributing the MSVC Runtime per [Microsoft Learn](/cpp/windows/deployment-in-visual-cpp).

## Selecting a Visual C++ Toolset

> By default, vcpkg will use the latest version of Visual Studio installed on the system for building code. To select a specific version, create a custom triplet or triplet overlay to set it.

Reviewer 2: ```suggestion
By default, vcpkg will use the latest version of Visual Studio installed on the system for building code. To select a specific version, create a custom triplet or triplet overlay to set [`VCPKG_PLATFORM_TOOLSET`](../triplets.md#vcpkg_platform_toolset).
```

Author: Fixed

For examples, this would force the use of the VS 2017 toolset.

```
set(VCPKG_PLATFORM_TOOLSET v141)
```

> See [Triplet files](../triplets.md) for more information.

Reviewer 2: ```suggestion
```

We can link to the relevant section directly above.

Author: Fixed

> ## C/C++ Runtime

Reviewer 2: ```suggestion
## C/C++ Runtime Compatibility
```

> Keep in mind that the Microsoft Visual C/C++ Runtime is 'forward binary compatible'. This means you can build code with VS 2015 Update 3, VS 2017, VS 2019, and/or VS 2022 and link it all together. The key requirement is that the LINK must be done against the *newest* toolset in the mix. See [Microsoft Learn](/cpp/porting/binary-compat-2015-2017).

Reviewer 2: ```suggestion
The Microsoft Visual C/C++ Runtime is 'forward binary compatible'. This means you can build code with VS 2015 Update 3, VS 2017, VS 2019, and/or VS 2022 and link it all together. The key requirement is that the LINK must be done against the *newest* toolset in the mix. See [Microsoft Learn](/cpp/porting/binary-compat-2015-2017).
```

Author: Fixed

## Known issues

* If you get the following link failure, it is because you are building x64 native code with VS 2019 or later, but linking it with the VS 2017 toolset & CRT combination. Per the notes above, you should use the newest compiler toolset for your final link. This *specific* issue can be worked around as well with ``/d2FH4-`` (v142 or later) to use the older x64 C++ exception implementation.

```
error LNK2001: unresolved external symbol ___CxxFrameHandler4
```

See the [Visual C++ Team Blog](https://devblogs.microsoft.com/cppblog/making-cpp-exception-handling-smaller-x64/) for more information.

* Full C++20 support requires VS 2019 (16.11.14 or later) and/or VS 2022 (17.2 or later). See the [Visual C++ Team Blog](https://devblogs.microsoft.com/cppblog/msvc-cpp20-and-the-std-cpp20-switch/).

> ## Maintainer notes

Reviewer 2: This is getting very off-topic for the vcpkg documentation -- at least as worded. This is not the right set of docs to teach generic C++ users about Windows.

However, if the following was intended to say "Consider creating a custom Windows triplet with settings X, Y, and Z if you care about them", that would be appropriate.

Author: Removed the Known Issues.

CMake projects for these triplets are built using ``CMAKE_SYSTEM_NAME`` set to "Windows".

> * "Just My Code" debugging can usually be disabled in a library to save code space.

Reviewer 2: ```suggestion
## Library author notes

* "Just My Code" debugging can usually be disabled in a library to save code space.
```

We typically reserve "maintainer" to refer to the owner of the port, which is different than the author of the library.
